forwardauth/handlers.go

```diff
@@ -0,0 +1,152 @@
+package forwardauth
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/gorilla/mux"
+
+	"github.com/pomerium/pomerium/internal/httputil"
+	"github.com/pomerium/pomerium/internal/sessions"
+	"github.com/pomerium/pomerium/internal/urlutil"
+	"github.com/pomerium/pomerium/pkg/cryptutil"
+)
+
+var registerHandler = map[string]func(*ForwardAuth, *mux.Router){
+	ProxyTypeNginx:   registerNginxHandlers,
+	ProxyTypeTraefik: registerTraefikHandlers,
+}
```

Comment:
Maybe we should change this to an interface in instead? Something like:

```go
type Proxy interface {
    RegisterHandlers(*ForwardAuth, *mux.Router)
}

var proxies = map[string]Proxy{
    ProxyTypeNginx: new(Nginx), 
    ProxyTypeTraefik: new(Traefik),
}
```

Or store an instance in the state and call it directly?

Reply:
@calebdoxsey I thought about that, too. But the fact is that if user wants to add new forward auth provider, they still need to modify code. I'm not sure the interface here has any advantage at all.

Reply:
Anyway, I change it to an interface, and detach related methods to given proxy type. Hope it's better. PTAL.

Reply:
I think you missed this comment because it was under the fold @cuonglm